### ISSUE TITLE
Reader skips over large block of messages after calling `seek`

### ISSUE THREAD

**Issue by gfredericks** [apache/pulsar#6847]
# Bug Description
In some difficult-to-reproduce circumstances, a reader will skip forward to later in the topic with no indication that anything abnormal happened.

# To Reproduce
1. Publish to a topic for a while; high volume might help
2. Choose two timestamps, `t1` and `t2`, with a decent window between them, that are both within the `publishTimestamp` history for the topic
3. Do the following dozens of times
    1. Create a reader and call `.seek(t1)`
    2. Read initial messages until you get one in between `t1` and `t2` (I would think the first message should qualify, but often it doesn't, and I don't know why)
    3. Read _and count_ messages until you see one with a `publishTimestamp > t2`
    4. Print/return the count
4. Compare the results across runs

# Expected behavior
The result should always be the same, assuming the publisher is setting `publishTimestamp` monotonically and has already passed `t2` at the beginning of the test.

# Additional context
I often see that during the run the producer makes at most one jump. Just recently I have noticed that the point it jumps to (the "jump target") increases over time, in a way that I think tracks the messages being published to the topic. I saw a very reliable increase until I stopped the producer, and then the increase stopped.

## Configurations I've Tested
- Broker versions 2.4.2, 2.5.1
- Producer version 2.3.0, 2.5.1
- Non-partitioned topics
- Batching enabled
- Python client, but mostly Java client
- Both reader and consumer

**Comment by sijie** [comment]
@gfredericks I moved the conversations here:

---

from @gfredericks [comment](https://github.com/streamnative/pulsar/issues/931#issuecomment-624080127)

Okay, I've figured out by analyzing the broker logs that this is is presumably due to the consumer_backlog_eviction policy:

21:41:51.542 [pulsar-backlog-quota-checker-26-1] INFO  org.apache.pulsar.broker.service.BacklogQuotaManager - Backlog quota exceeded for topic [persistent://public/default/tmp-integers]. Applying [consumer_backlog_eviction] policy
21:41:51.542 [pulsar-backlog-quota-checker-26-1] INFO  org.apache.bookkeeper.mledger.impl.ManagedCursorImpl - [public/default/persistent/tmp-integers] Skipping 12914 entries on cursor reader-ecf0525784
I don't understand all of the pieces here, but I would say that if this is the expected behavior it is rather unideal. My interpretation of a Reader is that it intends to read every message; if the system can't handle that for some reason or another, it would at least be useful to be able to detect that from the read side so the reader can retry or propagate the failure.

I'm going to keep researching the background issues so I can understand this better.

**Comment by sijie** [comment]
@gfredericks 

I see. The backlog quota policy is applied at the topic level. By default it is disabled. Did you happen to enable backlog quota policy? 

/cc @jiazhai @codelipenghui @gaoran10 in the conversation. So they can help you dig more into the problem.

**Comment by gfredericks** [comment]
The backlog quota is set to `consumer_backlog_eviction`, so I believe this is the explanation for what I'm seeing; it's possible that the resolution is just "remove the backlog quota", but I don't yet understand the implications of this to do it confidently.

Specifically:

If we intend to keep a very long history for the topic, is there any reason to have a backlog quota? I think we originally set it because we were worried about slow subscribers using a lot of resources, but do subscribers actually use much additional resources when there is already a long retention policy? Ideally a subscriber just consists of a few pointers into the data and the list of unacked message ids and perhaps some amount of data caching, but not an amount of resource usage linear in the messages between it and the end of the topic.

**Comment by gfredericks** [comment]
I noticed if I publish 10 million messages to a topic, meanwhile a slow subscriber consumes just a handful of them, the stats show something like

```
  "storageSize" : 1429259512,
  "backlogSize" : 1429122004,
      // for the subscription
      "msgBacklog" : 10393,
```

What do these numbers mean? Without subscribers the backlogSize is 0; when I add a subscriber at the beginning, it jumps up to 1.4G; does that mean the brokers or bookies literally copied all 1.4G of messages into a backlog? If not, why is the number so large? If so, what does the "10393" mean for the subscription?

**Comment by gfredericks** [comment]
Ultimately I want to know this:

Assuming I have infinite retention (i.e., with tiered storage), what bad thing can happen by not having a backlog quota? Will extra disk space get used? How much? In what situations?

**Comment by gfredericks** [comment]
After talking to Ming on slack, I'm pretty confident that I can safely remove the backlog quota.

However, I would say this isn't an ideal solution. Since readers (as opposed to consumers) intend to read _every_ message, and aren't necessarily trying to follow the end of the topic, I think it would make sense to have `consumer_backlog_eviction` and TTLs not apply to reader cursors, and if a reader is reading messages that become deleteable, the reader could have an exception thrown.

Silently skipping messages is a hard to detect and work around.

**Comment by gfredericks** [comment]
Ming judged this as a bug, saying that readers should never skip messages.

If that's the case, I'd be happy to try to contribute a fix, but would need to know what the correct behavior is. Presumably if readers don't skip messages, then there is a scenario where a reader has its messages fall off the retention cliff and get deleted, so the question is what should happen in that case. I think the reader should throw an exception in the client.

**Comment by sijie** [comment]
@gfredericks sorry for the late response.

I think the contract is the reader should read every message that is available in the storage. The `consumer_backlog_eviction` and `TTL` are the policies attached for quota management for topics in a namespace. These settings are how brokers manage the storage for topics that should be independent of how readers and consumers receiving messages. In other words, if a reader to read messages, people should make sure those storage quota policies don't apply to the topics.

With that being said, if people do configure those storage quota settings to the topics and a reader is reading those topics, the broker should detect that the data is cleaned up due to its storage quota policy and throw an exception to the client without silently dropping the messages.

So +1 to your last comment that we need to throw an exception to the clients. A pull request would work for adding this fix.

**Comment by gfredericks** [comment]
So I had been assuming that the bug only applied to readers, and that for consumers (non-readers) it's okay to skip messages, as I assume that's the whole point of `consumer_backlog_eviction`.

But in your message you didn't seem to be distinguishing, so just for clarity:

Should the exception we're talking about be thrown just for readers, or also for consumers?

**Comment by sijie** [comment]
Yes. Consumers are expected to skip messages in the context of "consumer_backlog_eviction". So the exception is thrown for readers.
